Add API Permission checking

suneettipirneni opened the pull request:
This adds an `guildRoles` and `guildUsers` field to `Command`.

### How is this different from `Command#permissions`?

`allowedRoles` and `allowedUsers` are a discord specific endpoint that prevents users from even initiating commands in the first place. For example if you're not allowed to run the `tester` command you'll see this:
<img width="235" alt="Screen Shot 2021-07-18 at 4 29 03 PM" src="https://user-images.githubusercontent.com/77477100/126081281-acd27039-463e-4ffb-b6a2-ec84cb21ec75.png">

It's grayed out and un-selectable. 

### So is the role checking in our `permissions` field useless now?

No, the role checking we crafted is dynamic by nature, it can change based off of input given to a certain commands. The discord API is static, once the permissions are set they don't change, unless the command itself is overwritten.

suneettipirneni committed:
Add API Permission checking

rob-3 commented:
This is a good idea, but it seems unclear to users whether to use the dynamic permission or regular for standard use cases. I would suggest renaming the permissions property to `dynamicPermissions` to make it clearer not to use it unless you need that.

suneettipirneni commented:
> This is a good idea, but it seems unclear to users whether to use the dynamic permission or regular for standard use cases. I would suggest renaming the permissions property to `dynamicPermissions` to make it clearer not to use it unless you need that.

How does just `Command#permissionHandler` sound?

rob-3 commented:
Works for me!

rob-3 commented:
Maybe `permissionsHandler` instead?

suneettipirneni commented:
> Maybe `permissionsHandler` instead? I'm not sure which sounds better

that works

suneettipirneni committed:
rename to `permissionsHandler`

rob-3 commented:
Oh wait I just realized the type is called `PermissionHandler`

suneettipirneni committed:
use singular.

rob-3 commented:
It looks like you used `guildRoles` and `guildChannels` instead of `allowedRoles` and `allowedChannels` in the code.

suneettipirneni commented:
> It looks like you used `guildRoles` and `guildChannels` instead of `allowedRoles` and `allowedChannels` in the code.

My bad, `guildRoles` and `guildChannels` were intentional.

rob-3 commented:
> > It looks like you used `guildRoles` and `guildChannels` instead of `allowedRoles` and `allowedChannels` in the code.
> 
> My bad, `guildRoles` and `guildChannels` were intentional.

Any particular reason for this? `allowedRoles` and `allowedChannels` seem more intuitive to me

suneettipirneni commented:
> > > It looks like you used `guildRoles` and `guildChannels` instead of `allowedRoles` and `allowedChannels` in the code.
> > 
> > 
> > My bad, `guildRoles` and `guildChannels` were intentional.
> 
> Any particular reason for this? `allowedRoles` and `allowedChannels` seem more intuitive to me

Yeah your right

suneettipirneni committed:
rename fields

rob-3 committed:
Refactor pushCommands()

I broke some functionality out to smaller functions and tried to
simplify the error handling logic.

rob-3 commented:
I refactored some things to reduce some duplication in the branches and simplify things. I wasn't sure about throwing if there wasn't a `GUILD_ID` or if `guild` didn't get initialized, but it seemed like the later code wouldn't work without this initialized.

rob-3 commented:
This is good to go as long as you like the refactoring changes.
